lib/private/converter.rb

```diff
@@ -52,23 +50,27 @@ def _convert(value, type)
         )
 
         if !true_idx.nil? && !false_idx.nil?
-          _convert_simple(value, T::Boolean)
+          _convert_simple(value, T::Boolean, raise_value_error)
         else
-          _convert(value, type.types[nil_idx == 0 ? 1 : 0])
+          _convert(value, type.types[nil_idx == 0 ? 1 : 0], raise_value_error)
         end
       elsif Object.const_defined?('T::Private::Types::TypeAlias') &&
             type.is_a?(T::Private::Types::TypeAlias)
-        _convert(value, type.aliased_type)
+        _convert(value, type.aliased_type, raise_value_error)
       elsif type < T::Struct
-        args = _build_args(value, type.props)
-        type.new(args)
+        args = _build_args(value, type, raise_value_error)
+        begin
+          type.new(args)
+        rescue
```

Comment:
Oh we rescue any error and raise as ShapeError? Why not just let it bubble through? (maybe you have a good reason here, I'm asking to understand the usecase) (You probably should acquire a practice of commenting on code that isn't obvious to people).

I thought ShapeError refers to errors related to the shape of the object (eg expect Hash but not a hash or expect Array but not an Array).

Also, it's likely a bad idea to catch any kind of error. You may end up catching errors not intended to be handled here. It's better to handle the exact error type you want to handle.

Reply:
The main motivation here is to see all errors we get when constructing a `T::Struct` as the arguments having a wrong shape. I guess this idea is debatable.

Reply:
That makes sense. Given that, I think you want to catch any error that's particularly about wrong "shape" (not TypeError caused by a member of the shape). If there is none at the moment, that's fine, we don't need to handle any error here.